helios-system-tests/src/main/java/com/spotify/helios/system/LabelTest.java

```diff
@@ -24,83 +24,122 @@
-       new Callable<Map<String, String>>() {
-         @Override
-         public Map<String, String> call()
-             throws Exception {
-           final Map<String, HostStatus> status = Json.read(
-               cli("hosts", testHost(), "--json"),
-               new TypeReference<Map<String, HostStatus>>() {
-               });
-           final Map<String, String>
-               labels =
-               status.get(testHost()).getLabels();
-           if (labels != null && !labels.isEmpty()) {
-             return labels;
-           } else {
-             return null;
-           }
-         }
-       });
-
-    assertThat(cli("hosts", "--labels", "role=foo"), containsString(testHost()));
-    assertThat(cli("hosts", "--labels", "xyz=123"), containsString(testHost()));
-    assertThat(cli("hosts", "--labels", "role=foo", "xyz=123"), containsString(testHost()));
-
-    assertThat(cli("hosts", "--labels", "role=doesnt_exist"), not(containsString(testHost())));
-    assertThat(cli("hosts", "--labels", "role=doesnt_exist", "xyz=123"),
-               not(containsString(testHost())));
+    // Wait for the agents to report labels
+    Polling.await(LONG_WAIT_SECONDS, SECONDS, hostLabels(ImmutableSet.of(testHost1, testHost2)));
+
+    // Test all these host selectors in one test method to reduce testing time.
+    // We can't start masters and agents in @BeforeClass because startDefaultMaster() isn't static
+    final String fooHosts = cli("hosts", "--labels", "role=foo");
+    assertThat(fooHosts, containsString(testHost1));
+    assertThat(fooHosts, not(containsString(testHost2)));
+
+    final String xyzHosts = cli("hosts", "--labels", "xyz=123");
+    assertThat(xyzHosts, containsStrings(ImmutableSet.of(testHost1, testHost2)));
+
+    final String fooAndXyzHosts = cli("hosts", "--labels", "role=foo", "xyz=123");
+    assertThat(fooAndXyzHosts, containsString(testHost1));
+    assertThat(fooAndXyzHosts, not(containsString(testHost2)));
+
+    final String notFooHosts = cli("hosts", "--labels", "role!=foo");
+    assertThat(notFooHosts, not(containsString(testHost1)));
+    assertThat(notFooHosts, containsString(testHost2));
+
+    final String inFooBarHosts = cli("hosts", "--labels", "role in (foo, bar)");
+    assertThat(inFooBarHosts, containsStrings(ImmutableSet.of(testHost1, testHost2)));
+
+    final String notInFooBarHosts = cli("hosts", "--labels", "role notin (foo, bar)");
+    assertThat(notInFooBarHosts, not(containsStrings(ImmutableSet.of(testHost1, testHost2))));
+
+    final String nonHosts = cli("hosts", "--labels", "role=doesnt_exist");
+    assertThat(nonHosts, not(containsStrings(ImmutableSet.of(testHost1, testHost2))));
+
+    final String nonHosts2 = cli("hosts", "--labels", "role=doesnt_exist", "xyz=123");
+    assertThat(nonHosts2, not(containsStrings(ImmutableSet.of(testHost1, testHost2))));
+  }
+
+  private CustomTypeSafeMatcher<String> containsStrings(final Set<String> substrings) {
```

Comment:
i think this could also be something like `allOf(containsString(".."), ..)` if you knew the set of strings at build-time